fireplace/dsl/random_picker.py

```diff
@@ -1,79 +1,96 @@
 import random
 from hearthstone.enums import CardType, Race, Rarity
 from .lazynum import LazyValue
+from copy import deepcopy
+from .. import cards
 
 
 class RandomCardPicker(LazyValue):
 	"""
 	Store filters and generate a random card matching the filters on pick()
+	Constructor takes a single global set of filters and uses a default weighting of 1
+	Additional weighted filter sets can be added with add(), these will be merged with the global filters
 	"""
-	def __init__(self, *args, **filters):
-		self.args = args
-		self.filters = filters
+	def __init__(self, **filters):
+		self.weights = []
+		self.weightedfilters = []
+		self.globalfilter = filters
 		self.count = 1
-		self._cards = None
-		self.lazy_filters = False
-		for v in filters.values():
-			if isinstance(v, LazyValue):
-				self.lazy_filters = True
-				break
 
 	def __repr__(self):
-		return "%s(%r)" % (self.__class__.__name__, self.filters)
+		return "%s(%r)" % (self.__class__.__name__, self.globalfilter)
 
+	# select number of cards to fetch
 	def __mul__(self, other):
-		ret = self.__class__(*self.args, **self.filters)
+		ret = deepcopy(self)
 		ret.count = other
 		return ret
 
-	@property
-	def cards(self):
-		if self._cards is None:
-			self._cards = self._filter_cards(self.filters)
-		return self._cards
+	# add a filter set
+	def add(self, weight, **filters):
```

Comment:
I don't really like that because of the extra arguments every Random* picker will have unless you specify weights=[,,,] as a keyword argument, but then you'll have to supply all the filters like that too.

The idea is to leave the DSL definitions alone and then in Discover simply take the supplied Random* picker and call eg. ".add(1, card_class=CardClass.NEUTRAL).add(4, card_class=<however we figure out the discover class to use>)". That way it is totally transparent and also flexible for the future.